quantum/quantum_read_write_mutex.h

```diff
@@ -129,59 +154,20 @@ class ReadWriteMutex
         /// @note Does not block.
         bool tryLock();
-
-        /// @brief Releases the read lock on the underlying mutex.
-        /// @note Also releases ownership of the underlying mutex.
-        void unlock();
         
-        /// @brief Release the associated mutex without unlocking it.
-        void release();
+        /// @brief Upgrade this reader to a writer.
+        /// @note Blocks until upgrade is performed.
+        /// @note The lock must already be owned prior to invoking this function.
+        void upgradeToWrite();
+        void upgradeToWrite(ICoroSync::Ptr sync);
         
-        /// @brief Indicates if this object owns the underlying mutex.
-        /// @return True if ownership is acquired.
-        bool ownsLock() const;
-    private:
-        ReadWriteMutex* _mutex{nullptr};
-        bool            _ownsLock{false};
-    };
-    
-    class WriteGuard
-    {
-    public:
-        /// @brief Construct this object and lock the passed-in mutex as a writer.
-        /// @param[in] lock ReadWriteMutex which protects a scope during the lifetime of the Guard.
-        /// @note Blocks the current thread until the mutex is acquired.
-        explicit WriteGuard(ReadWriteMutex& lock);
-        WriteGuard(ICoroSync::Ptr sync,
-                   ReadWriteMutex& lock);
-        
-        /// @brief Construct this object and tries to lock the passed-in mutex as a writer.
-        /// @param[in] lock ReadWriteMutex which protects a scope during the lifetime of the Guard.
-        /// @note Attempts to lock the mutex. Does not block.
-        WriteGuard(ReadWriteMutex& lock,
-                   ReadWriteMutex::TryToLock);
-        
-        /// @brief Construct this object and does not lock the mutex. Assumes the application
-        ///        already owns the write lock.
-        /// @param[in] lock ReadWriteMutex which protects a scope during the lifetime of the Guard.
-        /// @note: Does no block.
-        WriteGuard(ReadWriteMutex& lock,
-                   ReadWriteMutex::AdoptLock);
-        
-        /// @brief Destroy this object and unlock the underlying mutex if the object owns it.
-        ~WriteGuard();
-        
-        /// @brief Acquire the underlying mutex as a writer.
-        /// @note Blocks.
-        void lock();
-        void lock(ICoroSync::Ptr sync);
-        
-        /// @brief Try to acquire the underlying mutex as a writer.
-        /// @return True if mutex is locked, false otherwise.
+        /// @brief Try to upgrade this reader to a writer.
+        /// @return True is succeeded.
         /// @note Does not block.
-        bool tryLock();
+        /// @note The lock must already be owned prior to invoking this function.
+        bool tryUpgradeToWrite();
 
-        /// @brief Releases the write lock on the underlying mutex.
+        /// @brief Releases the read lock on the underlying mutex.
         /// @note Also releases ownership of the underlying mutex.
```

Comment:
This comment isn't quite right (I expect it's replicated on others); see comments above on destructor.